File: day07.js
const input = require("fs")
  .readFileSync("day07_input.txt")
  .toString()
  .split(/\r?\n/)

var pwd = ["/"]
const dictionary = {
  "/": {
    type: "dir",
    contents: [],
    size: 0,
  },
}

for (let i = 0; i < input.length; i++) {
  let line = input[i]
  let parent = pwd[pwd.length - 1]
  //console.log({ pwd })
  //command
  if (isCmd(line)) {
    let [$, cmd, arg] = line.split(" ")
    if (cmd === "cd") {
      //changes pwd
      handleCd(arg)
    } else if (cmd === "ls") {
      //do nothing
    }
  }
  //response to LS
  else {
    let [prefix, name] = line.split(" ")
    handleLsResponse(parent, prefix, name)
  }
}

Object.keys(dictionary).forEach((entryName) => {
  if (dictionary[entryName].type === "file") {
    addSizeToParent(entryName)
  }
})

let solution = 0
for (entry in dictionary) {
  //console.log(entry)
  if (dictionary[entry].type === "dir" && dictionary[entry].size < 100000) {
    solution += parseInt(dictionary[entry].size)
  }
}

console.log(`The solution is ${solution}`)

function addSizeToParent(fileName) {
  let queue = [dictionary[fileName].parent_dir]
  while (queue.length > 0) {
    let parent = queue.shift()
    //console.log({ parent })
    if (dictionary[parent]) {
      dictionary[parent].size += parseInt(dictionary[fileName].size)
      queue.push(dictionary[parent].parent_dir)
    }
  }
}

console.log(dictionary)

function isCmd(line) {
  return line.startsWith("$")
}

function handleCd(arg) {
  switch (arg) {
    case "/":
      //top dir
      pwd = ["/"]
      break
    case "..":
      //up one dir
      pwd.pop()
      break
    default:
      pwd.push(arg)
    //change dir
  }
  //console.log(pwd)
}

/*
const dictionary = {
  "/": {
    type: "dir",
    contents: ["a", "b.txt"],
    size: 0
  }
}
*/

function handleLsResponse(parent, prefix, name) {
  addChildrenToParent(parent, name)
  if (prefix === "dir") {
    addToDict("dir", name, parent)
  } else {
    addToDict("file", name, parent, prefix)
  }
}

function addChildrenToParent(parent, name) {
  if (dictionary[parent]) {
    if (dictionary[parent].contents.includes(name)) {
    } else {
      dictionary[parent].contents.push(name)
    }
  } else {
    dictionary[parent] = {
      type: "dir",
      contents: [name],
      size: 0,
    }
  }
}

function addToDict(type, name, parent, size) {
  if (dictionary[name]) {
  } else {
    if (type === "dir") {
      dictionary[name] = {
        type: "dir",
        contents: [],
        size: 0,
        parent_dir: parent,
      }
    } else {
      dictionary[name] = {
        type: "file",
        contents: [],
        size: parseInt(size),
        parent_dir: parent,
      }
    }
  }
}

File: day07.py
pwd = []
directory = {}


def is_command(str):
    return str.startswith("$")


def change_pwd(arg):
    global pwd
    if arg == "/":
        pwd = []
    elif arg == "..":
        pwd.pop()
    else:
        pwd.append(arg)


def generate_path(pwdArr, name):
    joinedArr = ("/").join(pwdArr)
    if joinedArr:
        return f"/{joinedArr}/{name}"
    else:
        return f"/{name}"


def add_entry_to_dict(type, path, size):
    global directory
    if not path in directory:
        if type == "dir":
            directory[path] = {
                "type": 'dir',
                "size": 0
            }
        else:
            directory[path] = {
                "type": "file",
                "size": int(size)
            }


def handle_ls_response(prefix, path):
    if prefix == "dir":
        add_entry_to_dict("dir", path)
    else:
        add_entry_to_dict("file", path, prefix)


def find_dir_size(dir):
    global directory
    dir_size = 0
    for item in directory:
        if item.startswith(dir) and directory[item].type == "file":
            dir_size += directory[item].size


with open('day07_input.txt') as f:
    terminal_list = f.read().splitlines()

# print(terminal_list)


for line in terminal_list:
    if is_command(line):
        [cmd_prefix, cmd, *arg] = line.split(" ")
        if (cmd == "cd"):
            change_pwd(arg)
        else:
            pass
    else:
        [prefix, name] = line.split(" ")
        path = generate_path(pwd, name)
        handle_ls_response(prefix, path)

solution = 0
for entry in directory:
    if directory[entry][type] == "dir":
        dir_size = find_dir_size(entry)
        if dir_size <= 100000:
            solution += dir_size

print(f"the solution is {solution}")
